### ISSUE TITLE
Ranger freezes when running a command and an image file is selected

### ISSUE THREAD

**Issue by Twix53791** [ranger/ranger#2615]
I noticed a strange behavior : 
when I run a :shell command from ranger and, unfortunately, an image file is currently selected in ranger, ranger freezes. It is not anymore responds to any command/shortcut, even the "reset" one, even the Q. I have to kill the window/process from outside.
When I say "selected", is not the selection <space> feature of ranger, but just when a file is currently under the "cursor", highlighted.

To reproduces : 
use any shell command like fzf, or ls (in ranger console: ':shell ls'), when a JPG or PNG file is selected.

It seems to happen only when image files (I tried jpg, png, NEF, ico) is highlighted, but NOT svg. Why? Ranger, with a svg file, display this message error : cannot identify image file '/home/test.svg'.
If I run the command with %p argument, I verified than the selected file path is well sent by ranger to the shell command even in the case of a svg file. 
So I guess ranger does something not explicit with the image file when it is currently selected and a shell command is launched, even completely unrelated with this file.
I don't know how can I enable/output a log in ranger to get a trace of what happening when it freezes?
Thanks to any hints !

**Comment by Twix53791** [comment]
Edit : it is only crash when the 'preview files' options is selected. I use kitty to preview images, maybe something related to that. I am gonna try without kitty. I am still demanding of tips to display a traceback of this bug, or enable a log file somewhere, to see a bit what is going on in the python script...

**Comment by markus-bauer** [comment]
Please specify what version you're using.
Can you try the master branch, this issue might be fixed already:
https://github.com/ranger/ranger/pull/1758

**Comment by Twix53791** [comment]
OK, I see it is a problem with kitty... Not so problematic so, as I can run ranger via another terminal in the worst case. I am gonna do that when I will have a bit of time. I have to clean everything as I hacked a bit ranger. I am gonna try the master branch yes.

**Comment by Twix53791** [comment]
I just update this issue : I just add to reinstall my all arch install, so it was a good occasion to test this problem in ranger. I confirm, from a fresh install of ranger and kitty, I got this bug straightway I activate the image preview with kitty (which seems to work, in appearance). I installed Pillow-9.0.1, ranger with pacman (ranger-1.9.3-4), and the kitty version is 0.25.0-1.

**Comment by fwicht** [comment]
I confirm having the same issue with kitty.

**Comment by markus-bauer** [comment]
Can either of you confirm that the version you're using has the patch I linked?

These seem to be the relevant changes:
https://github.com/ranger/ranger/pull/1758/files

And by master branch, I meant the git master branch.
You can try something like this:
```
git clone https://github.com/ranger/ranger/
cd ranger
python3 ./ranger.py
```

**Comment by Twix53791** [comment]
That's true than I didn't installed ranger via git. Once done, I confirm this change resolves the bug. Well done!
I confirm, also, than ueberzug is faster and above all smoother than kitty to preview images...

**Comment by cbreak-black** [comment]
I worked around this issue by adding a timeout to the display kitty reply checking code. It seems that in some situations, the reply gets lost or not sent by kitty in the first place.

```
diff --git a/ranger/ext/img_display.py b/ranger/ext/img_display.py
index 61e10f96..80f2314f 100644
--- a/ranger/ext/img_display.py
+++ b/ranger/ext/img_display.py
@@ -19,6 +19,7 @@ import imghdr
 import os
 import struct
 import sys
+import select
 import warnings
 import json
 import threading

@@ -676,30 +670,31 @@ class KittyImageDisplayer(ImageDisplayer, FileManagerAware):
         # catch kitty answer before the escape codes corrupt the console
         resp = b''
         while resp[-2:] != self.protocol_end:
-            resp += self.stdbin.read(1)
+            if select.select([self.stdbin], [], [], 4.0)[0]:
+                resp += self.stdbin.read(1)
+            else:
+                raise ImageDisplayError('kitty reply timeout'.format(resp))
         if b'OK' in resp:
             return
         else:
             raise ImageDisplayError('kitty replied "{r}"'.format(r=resp))
```

**Comment by toonn** [comment]
Fixed in #1758.

**Comment by pvskp** [comment]
I'm still facing this with Kitty 0.35.2 and 1.9.3 (both via pipx and OpenSUSE repository). 

Using ueberzug as the preview method is a workaround, tho.
